### PR TITLE
[DataSource] Add missing class property context

### LINKED ISSUE #2617: apollo-datasource: missing property context
According to this javascript tutorial
https://github.com/apollographql/fullstack-tutorial/blob/master/final/server/src/datasources/user.js

The following file is missing the property `context`
https://github.com/apollographql/apollo-server/blob/master/packages/apollo-datasource/src/index.ts

Here's my workaround, but it would be nice if someone fixed it :)
```
// types/apollo-datasource/index.d.ts

import { DataSource } from 'apollo-datasource'

declare module 'apollo-datasource' {
  interface DataSource<TContext = any> {
    context: TContext
  }
}
```

### PR DESCRIPTION
**Pull request by melkir**
According to the JS tutorial, we should be able to access to the `context` property inside a datasource https://github.com/apollographql/fullstack-tutorial/blob/master/final/server/src/datasources/user.js

So here ya go ;)

Here's the small use-case I have used (tested with TS 3.5.2)
```
export class TestAPI extends DataSource<any> {
  initialize(config: DataSourceConfig<any>) {
    this.context = config.context
  }
}
```

Close #2617

### CODE REVIEW

**Comment by abernix** [comment]
@martijnwalraven 	Do you have any thoughts on this?  We currently only define the `context` property on `RESTDataSource`:

https://github.com/apollographql/apollo-server/blob/aa3e23de9fbcde41a1bd3cca9c2eacf734bae582/packages/apollo-datasource-rest/src/RESTDataSource.ts#L49

Though the tutorial does give the impression that it would be set here.  Did you want to keep the `context` property out of the base implementation for a particular reason?

File: packages/apollo-datasource/src/index.ts [hunk 1]
```diff
@@ -6,5 +6,6 @@ export interface DataSourceConfig<TContext> {
 }
 
 export abstract class DataSource<TContext = any> {
+  context!: TContext;
```
Comment (abernix) [inline packages/apollo-datasource/src/index.ts]: Would we want this to be `!`'d?  Seems plausible that current implementations don't have it or might not have it?  I think it definitely makes sense to have it non-null asserted in `RESTDataSource` where we can be certain that we'll call it, but this seems a bit different.

Granted, it's just a typing change, so not a runtime disaster, but just curious what the thinking was here.

**Comment by abernix** [comment]
Also, I see @martijnwalraven's comment on https://github.com/apollographql/apollo-server/issues/2617 now, so I take back my questioning of him. :)

**Comment by melkir** [comment]
Closing as this is a staling issue.

**Pull request closed without merging**